### PR TITLE
Fast loading of internal clients and connections from a configuration file

### PR DESCRIPTION
**Pull request by twischer-adit**
This patch introduces a very fast way to load internal clients and connections on start up of the JACK daemon.
It is required on our systems to perform a very fast system start up.

An example configuration file looks like the following:
l inprocess1 inprocess
load amp1 jalv http://lv2plug.in/plugins/eg-amp
c amp:out system:playback_1

An example command line could look like the following:
$ jackd -C /etc/jack_paths.conf -d alsa

### CODE REVIEW

File: common/JackLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        jack_error("could not load %s, status = 0x%2.0x",
+                 client_name.c_str(), status);
+        return;
+    }
+    if (status & JackNameNotUnique) {
+        jack_error("Internal client name `%s' not unique",
+                 client_name.c_str());
+        return;
+    }
+
```
Comment (kmatheussen) [inline common/JackLoader.cpp]: Why not catch all situations where status != 0?

Also, doesn't fServer->InternalClientLoad1 always return false when status != 0?

File: common/Jackdmp.cpp [hunk 2]
```diff
@@ -202,6 +202,7 @@ static void usage(FILE* file, jackctl_server_t *server, bool full = true)
             "               [ --port-max OR -p maximum-number-of-ports]\n"
             "               [ --slave-backend OR -X slave-backend-name ]\n"
             "               [ --internal-client OR -I internal-client-name ]\n"
+            "               [ -C configuration-file ]\n"
```
Comment (kmatheussen) [inline common/Jackdmp.cpp]: All the other options have long and short versions. Maybe the -C option also should have a long version.

File: common/JackLoader.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    int linenr = -1;
+    while (std::getline(infile, line))
+    {
+        linenr++;
+
+        std::istringstream iss(line);
+
+        char command = 0;
+        if ( !(iss >> command) ) {
+            /* ignoring empty line */
```
Comment (kmatheussen) [inline common/JackLoader.cpp]: Does it also ignore lines with whitespace only?

**Comment by kmatheussen** [comment]
Would it make sense to use "Load" instead of "l", and "Connect" instead of "c"? It's a little bit hard to see those small "l" and "c" characters.

**Comment by kmatheussen** [comment]
Maybe a more specific name than "Configuration" should be used. It could be confused with the .jackd configuration file.

**Comment by twischer-adit** [comment]
@kmatheussen 
> Would it make sense to use "Load" instead of "l", and "Connect" instead of "c"? It's a little bit hard to see those small "l" and "c" characters.

I would like to have the commands as short as possible because the arguments of the internal clients can be very long. But it would be fine for me to support `l`, `c`AND `load`, `connect` as valid commands.


> Maybe a more specific name than "Configuration" should be used. It could be confused with the .jackd configuration file.

I think you mean the `.jackdrc`, `/etc/jackdrc` and `/etc/jackd.conf` which can be used to define the default arguments list of the JACK daemon, right?
Some example names which came into my mind:
- audio-path-configuration
- audio-route-configuration
Have you possibly better name?

File: common/Jackdmp.cpp (hunk 2 above)
Comment (twischer-adit) [inline common/Jackdmp.cpp]: What do you think about `audio-path-conf`?

File: common/JackLoader.cpp (hunk 1 above)
Comment (twischer-adit) [inline common/JackLoader.cpp]: I have verified it. The following call stack is used.
```
JackServer::InternalClientLoad1()
JackServer::InternalClientLoadAux()
JackServer::InternalClientLoad1()
JackInternalClient::Open()
JackInternalClientChannel::ClientCheck()
JackEngine::ClientCheck()
```
All of these functions are returning a negative value when `status` was changed to non-zero.
Therefore I only evaluate the `status` to find the reason for a failure, now.

**Comment by twischer-adit** [comment]
@kmatheussen 
I could not found your comment anymore. Possible you have already found the answer.
> Does it also ignore lines with whitespace only?

I have tested it with spaces and tabs and it is ignoring the whole line without any error messages.

**Comment by kmatheussen** [comment]
First, I forgot to tell write I'm not a maintainer. I just had some time and looked quickly over some pull requests. But I would think/hope @falkTX would have pointed out the same things.

> I would like to have the commands as short as possible because the arguments of the internal clients can be very long. But it would be fine for me to support l, cAND load, connect as valid commands.

Yeah, it was just a suggestion. There might be other pros and cons are here.

> I think you mean the .jackdrc, /etc/jackdrc and /etc/jackd.conf which can be used to define the default arguments list of the JACK daemon, right?

Yes.


>Some example names which came into my mind:
>
>    audio-path-configuration
>    audio-route-configuration
> 
> Have you possibly better name?

Yes, both of these sound better.


> I could not found your comment anymore.

Outdated comments are hidden when you force-push. Search for "Show outdated".

File: common/Jackdmp.cpp (hunk 2 above)
Comment (falkTX) [inline common/Jackdmp.cpp]: Configuration is really a bad name here.
I'm not a native english speaker, so thinking of names is something I know I'm bad at...

Comment (twischer-adit) [inline common/Jackdmp.cpp]: I changed it to `audio-path-conf`.

**Comment by falkTX** [comment]
I'd like to merge this, but I still don't agree with the names used here.
It's not an audio configuration, but a list of clients and connections..
JackLoader as class name is very generic (I know the current ones in the codebase are not that better..)

@twischer-adit do you mind if I import your changes manually, adjusting just the overall names used?

**Comment by twischer-adit** [comment]
@falkTX 
> do you mind if I import your changes manually, adjusting just the overall names used?

I am fine with it. Feel free to change the names.
Give me a ping if you need any action from me.

**Comment by falkTX** [comment]
ok, I've commited your changes into a new `internal-session-file` branch.
take a look and let me know what you think.
I haven't tried this just yet, will do it soon.

**Comment by twischer-adit** [comment]
I am fine with the changes of https://github.com/jackaudio/jack2/commit/2ad48419a7012b35e5e394530638686c825be4ba.

    Acked-by: Timo Wischer <twischer@de.adit-jv.com>

**Comment by falkTX** [comment]
Alright, merged into master just now.
Thanks for the initial patch, it's quite a handy feature!

**Pull request closed without merging**